tests/unit/test__helpers.py

```diff
@@ -231,6 +262,33 @@ def test_retry_exceeds_max_cumulative(self, randint_mock, sleep_mock):
         sleep_mock.assert_any_call(32.25)
         sleep_mock.assert_any_call(64.125)
 
+    @mock.patch(u"time.sleep")
+    @mock.patch(u"random.randint")
+    def test_retry_exceeded_reraises_connection_error(self, randint_mock, sleep_mock):
+        randint_mock.side_effect = [875, 0, 375, 500, 500, 250, 125]
+
+        responses = [requests.ConnectionError] * 8
+        func = mock.Mock(side_effect=responses, spec=[])
+
+        retry_strategy = common.RetryStrategy(max_cumulative_retry=100.0)
+        with pytest.raises(requests.ConnectionError):
+            _helpers.wait_and_retry(func, _get_status_code, retry_strategy)
+
+        assert func.call_count == 8
+        assert func.mock_calls == [mock.call()] * 8
+
+        assert randint_mock.call_count == 7
+        assert randint_mock.mock_calls == [mock.call(0, 1000)] * 7
+
+        assert sleep_mock.call_count == 7
+        sleep_mock.assert_any_call(1.875)
```

Comment:
Where do these numbers come from?

Reply:
This is just copied from the above functional test, but the numbers are the expected exponential backoff times plus the randint_mock.side_effect times above.